core/src/main/scala/org/apache/spark/CacheRecoveryManager.scala

```diff
@@ -0,0 +1,201 @@
+    val execsToShutDown = execIds.toSet
+    // Memory Status is a map of executor Id to a tuple of Max Memory and remaining memory on that
+    // executor.
+    val futureMemStatusByBlockManager =
+        blockManagerMasterEndpoint.ask[Map[BlockManagerId, (Long, Long)]](GetMemoryStatus)
+
+    val futureMemStatusByExecutor = futureMemStatusByBlockManager.map { memStat =>
+      memStat.map { case (blockManagerId, mem) => blockManagerId.executorId -> mem }
+    }
+
+    futureMemStatusByExecutor.map { memStatusByExecutor =>
+      val (expiringMemStatus, remainingMemStatus) = memStatusByExecutor.partition {
+        case (execId, _) => execsToShutDown.contains(execId)
+      }
+      val freeMemOnRemaining = remainingMemStatus.values.map(_._2).sum
+
+      // The used mem on each executor sorted from least used mem to greatest
+      val executorAndUsedMem: Seq[(String, Long)] =
+        expiringMemStatus.map { case (execId, (maxMem, remainingMem)) =>
+          val usedMem = maxMem - remainingMem
+          execId -> usedMem
+        }.toSeq.sortBy { case (_, usedMem) => usedMem }
+
+      executorAndUsedMem
+        .scan(("start", freeMemOnRemaining)) {
+          case ((_, freeMem), (execId, usedMem)) => (execId, freeMem - usedMem)
+        }
+        .drop(1)
+        .filter { case (_, freeMem) => freeMem > 0 }
+        .map(_._1)
+    }
+  }
+
+  /**
+   * Given a value and a timeout in seconds, complete the future with the value when time is up.
+   *
+   * @param value The value to be returned after timeout period
+   * @param seconds the number of seconds to wait
+   * @return a future that will hold the value given after a timeout
+   */
+  private def returnAfterTimeout[T](value: T, seconds: Long): Future[T] = {
+    val p = Promise[T]()
+    val runnable = new Runnable {
+      def run(): Unit = { p.success(value) }
+    }
+    scheduler.schedule(runnable, seconds, TimeUnit.SECONDS)
+    p.future
+  }
+
+  /**
+   * Recover cached RDD blocks off of an executor until there are no more, or until
+   * there is an error
+   *
+   * @param execId the id of the executor to be killed
+   * @return a Future of Unit that will complete once all blocks have been replicated.
+   */
+  private def replicateUntilDone(execId: String): Future[KillReason] = {
+    recoverLatestBlock(execId).flatMap { moreBlocks =>
+      if (moreBlocks) replicateUntilDone(execId) else Future.successful(DoneRecovering)
+    }
+  }
+
+  /**
+   * Ask the BlockManagerMaster to replicate the latest cached rdd block off of this executor on to
+   * a surviving executor, and then remove the block from this executor
+   *
+   * @param execId the executor to recover a block from
+   * @return A future that will hold true if a block was recovered, false otherwise.
+   */
+  private def recoverLatestBlock(execId: String): Future[Boolean] = {
+    blockManagerMasterEndpoint
+      .ask[Boolean](RecoverLatestRDDBlock(execId, recoveringExecutors.asMap.keySet.asScala.toSeq))
```

Comment:
so the reason you do this one block at a time is because you update `recoveringExecutors` between each call, right?  To help avoid replicating to another executor which will start draining its blocks after this starts replicating the first block, but before it starts replicating later blocks?

if so, that explanation should go in a comment.